feat(react): handle cleanup of DOM feature app on it's removal

msaktor opened the pull request:
Hi! First thanks a lot for this amazing library. It allowed us to greatly scale frontend development at Jamf.

We would like to suggest a possibility to do a cleanup when `feature-app-container` is disconnected from the DOM, in order to avoid memory leaks.

Here I present a suggestion that uses a return function in `attachTo` in a similar fashion how "useEffect" works in React. The benefit is access to variables declared inside the `attachTo` scope.

Another solution could be a `detachFrom` function that can be an optional fn on `DomFeatureApp` interface.

Example for React:
```tsx
const featureAppDefinition: FeatureAppDefinition<DomFeatureApp> = {
  dependencies: {
    featureServices: {},
  },
  create: (env: FeatureAppEnvironment<FeatureServices, Config>) => {
    return {
      attachTo(element: HTMLElement): void {
        reactRoot = createRoot(element);
        reactRoot.render(<div>Hello world!</div>);

        return () => {
          reactRoot.unmount();
        };
      },
    };
  },
};
```

Or Vue:
```ts
const featureAppDefinition: FeatureAppDefinition<DomFeatureApp> = {
  dependencies: {
    featureServices: {},
  },
  create: (env: FeatureAppEnvironment<FeatureServices, Config>) => {
    return {
      attachTo(element: HTMLElement): void {
        const app = Vue.createApp({
          template: '<div>Hello world!</div>',
        });
        app.mount('#app');

        return () => {
           app.unmount();
        };
      },
    };
  },
};
```

NSteinhoff commented:
Hi @msaktor, thank you for your contribution! 

I think this is a good improvement. I also agree with your reasoning to prefer a cleanup callback à la `useEffect` over a `detachFrom` method.

---

One issue I see, is that a DOM feature app can also be rendered by a React `FeatureAppContainer`, which would currently not run the cleanup.

https://github.com/feature-hub/feature-hub/blob/11490453001ea2628ecb9aa291d494618f429305/packages/react/src/feature-app-container.tsx#L39

https://github.com/feature-hub/feature-hub/blob/11490453001ea2628ecb9aa291d494618f429305/packages/react/src/internal/internal-feature-app-container.tsx#L326

I think we should ensure consistent behavior of DOM feature apps for both DOM and React containers.

---

We currently have limited capacity to invest into active feature-hub development. You could help this along greatly by providing the appropriate documentation/demo/test updates. We'll try to get to it as soon as we are able.

Thanks,
Niko

msaktor committed:
feat(dom): add ability to do cleanup when app is removed

msaktor committed:
feat(react): handle cleanup of DOM feature app on it's removal

msaktor commented:
Hi Niko @NSteinhoff !

I've updated the PR with tests and examples, can you have another look please?

Thanks!
Marek

NSteinhoff reviewed (dismissed):
Hi @msaktor, thanks for the update! Looks great 👍 

We need to investigate and fix the failing builds. It's possible that they are not related to your changes, but I'm not yet familiar enough with the CI setup to be sure. Are you blocked by this?

NSteinhoff committed:
fix(dom,react): remove 'void' from union type

This is breaking on a type level (depending on the compiler settings and
type annotations of user-level code), but I don't think that it warrants
a major (or even minor) level version bump.

> The void type means “nothing” or that a function does not return any
> value, in contra with implicit undefined type which means that a
> function returns a value undefined. So “nothing” cannot be mixed with
> any other types. If you need this - use undefined type instead.

https://palantir.github.io/tslint/rules/invalid-void/

NSteinhoff reviewed (approved):
Hi @msaktor, I took the liberty to update the PR to meet the linting requirements. It's now good to go and I will merge it shortly. Thanks again for your contribution! 🥇 

The coverage action fails for PRs from forks, because it does not have the necessary rights to post the report. Good to know and we'll probably want to address it, but we can ignore it for this PR.

msaktor commented:
Thanks a lot for merging this Niko! @NSteinhoff 

I see that npm still has the latest version 3.7.0 and not 3.8.0, are you planning to publish the new version soon?

NSteinhoff commented:
Hi @msaktor, the `@feature-hub/react` and `@feature-hub/dom` packages should be at 3.8.0:

https://www.npmjs.com/search?q=feature-hub

msaktor commented:
my bad, I was looking at @feature-hub/core 🤦 
thanks!
